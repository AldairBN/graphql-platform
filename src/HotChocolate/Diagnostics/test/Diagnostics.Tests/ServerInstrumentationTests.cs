using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.TestHost;
using Microsoft.Extensions.DependencyInjection;
using HotChocolate.AspNetCore.Utilities;
using Snapshooter.Xunit;
using Xunit;
using static HotChocolate.Diagnostics.ActivityTestHelper;
using System;
using System.Net.Http;

namespace HotChocolate.Diagnostics;

public class ServerInstrumentationTests : ServerTestBase
{
    public ServerInstrumentationTests(TestServerFactory serverFactory)
        : base(serverFactory)
    {
    }

    [Fact]
    public async Task Http_Post_SingleRequest_GetHeroName()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer();

            // act
            ClientQueryResult result =
                await server.PostAsync(new ClientQueryRequest
                {
                    Query = @"
                    {
                        hero {
                            name
                        }
                    }"
                });

            // assert
            activities.MatchSnapshot();
        }
    }

    [Fact]
    public async Task Http_Get_SingleRequest_GetHeroName()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer();

            // act
            ClientQueryResult result =
                await server.GetAsync(new ClientQueryRequest
                {
                    Query = @"
                    {
                        hero {
                            name
                        }
                    }"
                });

            // assert
            activities.MatchSnapshot();
        }
    }

    [Fact]
    public async Task Http_Post_variables_are_not_automatically_added_to_activities()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer();

            // act
            ClientQueryResult result =
                await server.PostAsync(new ClientQueryRequest
                {
                    Query = @"
                    query ($episode: Episode!) {
                        hero(episode: $episode) {
                            name
                        }
                    }",
                    Variables = new Dictionary<string, object> { { "episode", "NEW_HOPE" } }
                });

            // assert
            activities.MatchSnapshot();
        }
    }

    [Fact]
    public async Task Http_Post_add_variables_to_http_activity()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer(
                o => o.RequestDetails = RequestDetails.Default | RequestDetails.Variables);

            // act
            ClientQueryResult result =
                await server.PostAsync(new ClientQueryRequest
                {
                    Query = @"
                    query ($episode: Episode!) {
                        hero(episode: $episode) {
                            name
                        }
                    }",
                    Variables = new Dictionary<string, object> { { "episode", "NEW_HOPE" } }
                });

            // assert
            activities.MatchSnapshot();
        }
    }

    [Fact]
    public async Task Http_Post_add_query_to_http_activity()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer(
                o => o.RequestDetails = RequestDetails.Default | RequestDetails.Query);

            // act
            ClientQueryResult result =
                await server.PostAsync(new ClientQueryRequest
                {
                    Query = @"
                    query ($episode: Episode!) {
                        hero(episode: $episode) {
                            name
                        }
                    }",
                    Variables = new Dictionary<string, object> { { "episode", "NEW_HOPE" } }
                });

            // assert
            activities.MatchSnapshot();
        }
    }

    [Fact]
    public async Task Http_Post_with_extensions_map()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer();

            // act
            ClientQueryResult result =
                await server.PostAsync(new ClientQueryRequest
                {
                    Query = @"
                    query ($episode: Episode!) {
                        hero(episode: $episode) {
                            name
                        }
                    }",
                    Variables = new Dictionary<string, object> { { "episode", "NEW_HOPE" } },
                    Extensions = new Dictionary<string, object> { { "test", "abc" } }
                });

            // assert
            activities.MatchSnapshot();
        }
    }

    [Fact]
    public async Task Http_Get_SDL_download()
    {
        using (CaptureActivities(out var activities))
        {
            // arrange
            using TestServer server = CreateInstrumentedServer();
            var url = TestServerExtensions.CreateUrl("/graphql?sdl");
            var request = new HttpRequestMessage(HttpMethod.Get, url);

            // act
            HttpResponseMessage response = await server.CreateClient().SendAsync(request);

            // assert
            await response.Content.ReadAsStringAsync();
         
            // assert
            activities.MatchSnapshot();
        }
    }

    private TestServer CreateInstrumentedServer(
        Action<InstrumentationOptions>? options = default)
        => CreateStarWarsServer(
                configureServices: services =>
                    services
                        .AddGraphQLServer()
                        .AddInstrumentation(options));
}
