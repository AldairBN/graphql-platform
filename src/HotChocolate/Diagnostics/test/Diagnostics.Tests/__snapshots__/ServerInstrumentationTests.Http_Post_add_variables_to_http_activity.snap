{
  "activities": [
    {
      "OperationName": "ExecuteHttpRequest",
      "DisplayName": "GRAPHQL HTTP POST",
      "Status": "Unset",
      "tags": [
        {
          "Key": "graphql.schema.name",
          "Value": "_Default"
        },
        {
          "Key": "graphql.http.request.type",
          "Value": "single"
        },
        {
          "Key": "graphql.http.request.query.id",
          "Value": "EnTNrU/qlue0XYlwvGskqA=="
        },
        {
          "Key": "graphql.http.request.query.hash",
          "Value": "EnTNrU/qlue0XYlwvGskqA=="
        },
        {
          "Key": "graphql.http.request.variables",
          "Value": "{ episode: \"NEW_HOPE\" }"
        }
      ],
      "event": [],
      "activities": [
        {
          "OperationName": "ParseHttpRequest",
          "DisplayName": "ParseHttpRequest",
          "Status": "Unset",
          "tags": [],
          "event": []
        },
        {
          "OperationName": "ExecuteRequest",
          "DisplayName": "Execute Request",
          "Status": "Unset",
          "tags": [
            {
              "Key": "graphql.document.id",
              "Value": "EnTNrU/qlue0XYlwvGskqA=="
            },
            {
              "Key": "graphql.operation.id",
              "Value": "EnTNrU/qlue0XYlwvGskqA=="
            }
          ],
          "event": [
            {
              "Name": "AddedOperationToCache",
              "Tags": {}
            },
            {
              "Name": "AddedDocumentToCache",
              "Tags": {}
            }
          ],
          "activities": [
            {
              "OperationName": "ValidateDocument",
              "DisplayName": "Validate Document",
              "Status": "Ok",
              "tags": [
                {
                  "Key": "graphql.document.id",
                  "Value": "EnTNrU/qlue0XYlwvGskqA=="
                }
              ],
              "event": []
            },
            {
              "OperationName": "CompileOperation",
              "DisplayName": "Compile Operation",
              "Status": "Unset",
              "tags": [],
              "event": []
            },
            {
              "OperationName": "CoerceVariables",
              "DisplayName": "Coerce Variable",
              "Status": "Unset",
              "tags": [],
              "event": []
            },
            {
              "OperationName": "BuildQueryPlan",
              "DisplayName": "Build Query Plan",
              "Status": "Unset",
              "tags": [],
              "event": []
            },
            {
              "OperationName": "ExecuteOperation",
              "DisplayName": "ExecuteOperation",
              "Status": "Unset",
              "tags": [],
              "event": [],
              "activities": [
                {
                  "OperationName": "ResolveFieldValue",
                  "DisplayName": "/hero",
                  "Status": "Ok",
                  "tags": [
                    {
                      "Key": "graphql.selection.name",
                      "Value": "hero"
                    },
                    {
                      "Key": "graphql.selection.parentType",
                      "Value": "Query"
                    },
                    {
                      "Key": "graphql.selection.coordinate",
                      "Value": "Query.hero"
                    },
                    {
                      "Key": "graphql.selection.type",
                      "Value": "Character"
                    },
                    {
                      "Key": "graphql.selection.path",
                      "Value": "/hero"
                    },
                    {
                      "Key": "graphql.selection.hierarchy",
                      "Value": "/hero"
                    }
                  ],
                  "event": []
                }
              ]
            }
          ]
        },
        {
          "OperationName": "FormatHttpResponse",
          "DisplayName": "FormatHttpResponse",
          "Status": "Unset",
          "tags": [],
          "event": []
        }
      ]
    }
  ]
}
